File: expression_evaluator.cpp
#include <iostream>
#include <stack>
#include <string>
#include <exception>
#include <cassert>

using namespace std;

class Error : public exception
{
public:
    Error(string const& desc = "Unknown Error") throw()
        :m_desc(desc)
    {}
    virtual const char* what() const throw()
    {
        return m_desc.c_str();
    }

    virtual ~Error() throw()
    {}

private:
    string m_desc;
};

template <typename T>
void evaluate(stack<T>& operands, stack<char>& operators) {

    T a = operands.top();
    operands.pop();
    T b = operands.top();
    operands.pop();

    switch (operators.top()) {

    case '+':
        operands.push(b + a);
        break;
    case '-':
        operands.push(b - a);
        break;
    case '*':
        operands.push(b * a);
        break;
    case '/':
        if (a == 0) throw Error("ZeroDivision");
        operands.push(b / a);
        break;
    default:
        throw Error("Invalid Operator");
        break;

    }

}

int precedence(char o) {
    if (o == '+' || o == '-') return 0;
    else if (o == '*' || o == '/') return 1;
    // case: math functions as the pow or ^
    else throw Error("Invalid Operator");
}

template <typename T>
T eval(string const& exp) {

    stack<T> operands;
    stack<char> operators;

    // To track the last element
    char last = '0';
    int open_p = 0;

    for (unsigned i = 0; i < exp.length(); i++) {
        
        // case: operand
        if (exp[i] >= '0' && exp[i] <= '9') {
            // test for duplicated operands
            if (last == 'N') throw Error("Invalid Expression::Duplicated Operand");

            T x = 0;
            unsigned unit;
            do {
                unit = exp[i] - '0';
                x = 10 * x + unit;
                i++;
                if (i == exp.length()) break;
            } while (exp[i] >= '0' && exp[i] <= '9');

            if (exp[i] == '.') {
                i++;
                unsigned puiss = 10;
                while (exp[i] >= '0' && exp[i] <= '9')
                {
                    unit = exp[i] - '0';
                    x += (T) unit / puiss;
                    puiss *= 10;
                    i++;
                    if (i == exp.length()) break;
                }

            }

            operands.push(x);
            last = 'N';

            // to reset to the current place after incrementation
            i--;
        }

        // case: operator
        else if (exp[i] == '+' || exp[i] == '-' || exp[i] == '*' || exp[i] == '/') {
            
            // test for duplicated operators
            if (last == 'O') throw Error("Invalid Expression::Duplicated Operator");

            if (operators.empty() || operators.top() == '(') operators.push(exp[i]);
            else if (precedence(exp[i]) > precedence(operators.top())) operators.push(exp[i]);
            else {

                do {

                    evaluate(operands, operators);
                    operators.pop();

                } while (!operators.empty() && !(operators.top() == '('));

                operators.push(exp[i]);
            }
            
            last = 'O';

        }

        // case: open parentheses
        else if (exp[i] == '(') {
            operators.push('(');
            open_p++;
        }

        // case: close parantheses
        else if (exp[i] == ')') {
            if (open_p < 1) throw Error("Invalid Expression::Unmatched ')'");

            do {

                evaluate(operands, operators);
                operators.pop();

            } while (operators.top() != '(');

            operators.pop();
            open_p--;

        }

        // case: space
        else if (exp[i] == ' ') continue;

        // case: Invalid Expression
        else throw Error("Invalid Expression::Unknown Character");
        
    }


    // final calculations
    if (!operators.empty()) {

        do {

            evaluate(operands, operators);
            operators.pop();

        } while (!operators.empty());

    }

    return operands.top();
}

int main()
{
    
    try {
        assert(eval<int>("14 * 5 + (55 - 20) - 9 / 8") == 14 * 5 + (55 - 20) - 9 / 8);

        // Testing errors:
        // cout << eval<double>("14.05 /f 0.5 + (100 * 20) - 9 / 8") << endl;
        // cout << eval<double>("14..05 / 0.5 + (100 * 20) - 9 / 8") << endl;
        // cout << eval<double>("14.0.5 / 0.5 + (100 * 20) - 9 / 8") << endl;
        // cout << eval<double>("14.05 / / 0.5 + (100 * 20) - 9 / 8") << endl;
        // cout << eval<float>("14 20 / 5 + (100 * 20) - 9 / 8") << endl;
        // cout << eval<float>("14 / 5 + (100 * 20)) - 9 / 8") << endl;
        // cout << eval<int>("14 * 5 + (55 - 20) - 9 / 0");

        cout << eval<float>("14 / 5 + (100 * 20) - 9 / 8") << endl;
        cout << eval<int>("14 / 5 + (100 * 20) - 9 / 8") << endl;
        cout << "All tests have been passed successfully" << endl;

    }
    catch (std::exception const& e) {
        cerr << "Error: " << e.what();
    }
    
    return 0;
}

File: expression_evaluator.py
def evaluate(operands, operators):
    
    a = operands.pop()
    b = operands.pop()
    
    if operators[len(operators)-1] == '+':
        operands.append(b + a)
        return b + a
    elif operators[len(operators)-1] == '-':
        operands.append(b - a)
        return b - a
    elif operators[len(operators)-1] == '*':
        operands.append(b * a)
        return b * a
    elif operators[len(operators)-1] == '/':
        if a == 0: raise ZeroDivisionError
        operands.append(b / a)
        return b / a
        
    # handle default case
    else: raise ValueError


def precedence(operator):
    if operator == '+' or operator == '-':
        return 0
    elif operator == '*' or operator == '/':
        return 1
    # case: math functions
    else: raise ValueError


def eval_exp(exp):

    operands = []
    operators = []

    i = 0
    # to track the last element
    last = '0'
    # to track parenthesis
    open_p = 0
    
    while i < len(exp):

        # case: operand
        if exp[i] >= '0' and exp[i] <= '9':

            if last == 'N': raise SyntaxError
            
            x = 0
            while exp[i] >= '0' and exp[i] <= '9':
                x = 10 * x + int(exp[i])
                i += 1
                if i == len(exp): break
                
            if i < len(exp):
                if exp[i] == '.':
                    i += 1
                    puiss = 10
                    while exp[i] >= '0' and exp[i] <= '9':
                        x += int(exp[i]) / puiss
                        puiss *= 10
                        i += 1
                        if i == len(exp): break 

            operands.append(x)
            last = 'N'

            # to reset to the current place after incrementation
            i -= 1
            
        # case: operator
        elif exp[i] == '+' or exp[i] == '-' or exp[i] == '*' or exp[i] == '/':

            if last == 'O': raise SyntaxError

            if len(operators) == 0 or operators[len(operators)-1] == '(':
                operators.append(exp[i])
            elif precedence(exp[i]) > precedence(operators[len(operators)-1]):
                operators.append(exp[i])
            else:

                while len(operators) != 0 and operators[len(operators)-1] != '(':
                    evaluate(operands, operators)
                    operators.pop()
                    
                operators.append(exp[i])
            last = 'O'

        # case: open parentheses
        elif exp[i] == '(':
            operators.append('(')
            open_p += 1

        # case: close parantheses
        elif exp[i] == ')':

            if open_p < 1: raise SyntaxError

            while operators[len(operators)-1] != '(':

                evaluate(operands, operators)
                operators.pop()

            
            operators.pop()
            open_p -= 1

        # case: space
        elif exp[i] == ' ': pass

        # case: default
        else: raise SyntaxError
        
        # increment i inside the while loop
        i += 1
    
    # final calculations
    if len(operators) != 0:

        while len(operators) != 0:

            evaluate(operands, operators)
            operators.pop()

    
    return operands.pop()

#testing our function
#the following block is not imported, but run as script
# __name__ is special variable defined by the iterpreter before execution
if __name__ == "__main__":
    
    try:
        print('Enter Q to quit the test')
        while True:
            test = input('Enter expression to evaluate: ')
            if test == 'Q':
                break
            assert eval(test) == eval_exp(test)
            print(eval_exp(test))
        

    except SyntaxError:
        print("Invalid Expression")
    except ZeroDivisionError:
        print("Zero Division Error")
    except ValueError:
        print("Invalid Operator")
    except AssertionError:
        print("Assert Error")
    except:
        print("Error Occured")
    finally:
        print('Finished')

